src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java

```diff
@@ -49,33 +44,80 @@ public String description()
 			return Collections.emptyList();
 		}
 
-		if (log.isDebugEnabled()) {
-			log.debug("getInstances: service = {}", service.toString());
+		return getInstancesFromService(service);
+	}
+
+	@Override
+	public List<ServiceInstance> selectInstances(Map<String, String> match)
+	{
+		return selectInstances(match, Collections.emptyMap());
+	}
+
+	@Override
+	public List<ServiceInstance> selectInstances(Map<String, String> match, Map<String, String> doNotMatch)
+	{
+		ServiceList serviceList;
+		try {
+			serviceList = kubeClient.services().withLabels(match).withoutLabels(doNotMatch).list();
+		} catch(KubernetesClientException e) {
+			log.warn("getInstancesMatchingMetadata: failed to retrieve matching services: API call failed: {}",
+				e.getMessage());
+			return Collections.emptyList();
 		}
 
-		if (service.getSpec().getPorts().isEmpty()) {
-			log.error("getInstances: service '{}' has no ports", serviceId);
+		if (serviceList == null) {
+			log.warn("getInstancesMatchingMetadata: service list is null");
 			return Collections.emptyList();
 		}
 
-		List<ServicePort> servicePorts = service.getSpec().getPorts();
-		ServicePort svcPort = servicePorts.get(0); // By default, use the first port available
+		return getInstancesFromServiceList(serviceList.getItems());
+	}
+
+	private List<ServiceInstance> getInstancesFromService(Service service)
+	{
+		List<Service> serviceList = new ArrayList<>();
+		serviceList.add(service);
+		return getInstancesFromServiceList(serviceList);
+	}
 
-		// But if a port named "http" is available, use it instead
-		Optional<ServicePort> httpPort = servicePorts.stream()
-			.filter(s -> s.getName() != null && s.getName().equals("http")).findFirst();
-		if (httpPort.isPresent()) {
-			svcPort = httpPort.get();
+	private List<ServiceInstance> getInstancesFromServiceList(List<Service> services)
+	{
+		if (log.isDebugEnabled()) {
+			log.debug("services = {}", services.toString());
```

Comment:
More of a context in the debugging message might be handy, after we forget about this log line.